### PR TITLE
[RFC] Introduce prettier-ignore

### LINKED ISSUE #120: Add escape hatch
`prettier` at this point is still outputting invalid JavaScript or badly formatted output in many cases. This will go down as time goes but in the meantime, it would be great to be able to disable it for just an expression.

Proposal:

```js
// prettier-ignore-next-block
function a() { 
  iCanDoWhateverIdent
    (a, b,); }
```

and prettier will keep the next AST node as is.

### PR DESCRIPTION
**Pull request by vjeux**
This was surprisingly easy to write!

`prettier-ignore` will leave the expression the comment is attached to printed as it was before.

Fixes #120

### CODE REVIEW

File: tests/disable/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -0,0 +1,90 @@
+exports[`test disable.js 1`] = `
+"function a() {
+  // prettier-ignore-next-line
```
Comment (aaronyo) [inline tests/disable/__snapshots__/jsfmt.spec.js.snap]: Is the test producing a false positive somehow? It looks the the code is expecting "prettier-disable-next-line" rather than "...-ignore-..."

**Comment by aaronyo** [comment]
My $.02 on naming: maybe ```// prettier-disable-next-node``` or even just ```// prettier-disable-next```. If I'm understanding the code correctly, the scope of the escape hatch is the next AST node? I think it might actually confuse me a tad at first if it is "next-line" but it doesn't behave like eslint.

Regardless, I'll be happy to have this feature! Thank you!

**Comment by rtsao** [comment]
Istanbul (which is fairly common) uses `/* istanbul ignore next */` for skipping the next expression/block. Agreed that it seems confusing if `next-line` would actually ignore the next block/expression.

File: tests/disable/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (vjeux) [inline tests/disable/__snapshots__/jsfmt.spec.js.snap]: I first named it ignore and then looked at eslint where it's disable. I must have forgotten to change this here :)

**Comment by vjeux** [comment]
Thanks for the feedback, I think that prettier-ignore-next is the best choice.

**Comment by vjeux** [comment]
Done!

**Comment by jlongster** [comment]
Yeah I was going to suggest a simpler name as well. We could even do `// prettier-ignore` and it would just read like a leading comment. I always have to look up eslint's ignore names because they are always 3-5 words. The easier we can make it to remember the better.

I don't think we need to worry about a learning curve and stick to eslint's conventions. This is something people will learn once and it's ok.

It there any chance you all would be ok with just `// prettier-ignore`?

**Comment by jlongster** [comment]
I'm interested to see if this ever interacts badly with our formatting. There might be places where if you use this it will inject code that leaves it in a weird indentation state or something. But I think this is the right idea.

File: src/printer.js [hunk 2]
```diff
@@ -46,6 +46,17 @@ function genericPrint(path, options, printPath) {
...
+  // Escape hatch
+  if (node.comments &&
+      node.comments.length > 0 &&
+      node.comments[0].value.trim() === 'prettier-ignore-next' &&
+      node.comments[0].leading) {
+    return options.originalText.slice(
+      util.locStart(node),
+      util.locEnd(node)
+    );
+  }
```
Comment (jlongster) [inline src/printer.js]: Try adding some test cases where the expression is an argument to function, a property of an object, etc. Those interactions are where things might get weird.

The problem is the above code will naively output `\n`'s and will mess up the measurement system. We might need split the content on newline and insert our own hardlines to make sure it never gets messed up. We might should add a `breakParent` just to make sure it always breaks up the expression too.

**Comment by vjeux** [comment]
`// prettier-ignore` works for me. I can even remove the `leading` check so you can use it as a trailing comment if you want to.

**Comment by aaronyo** [comment]
+1 for ```// prettier-ignore```

**Comment by jlongster** [comment]
> I can even remove the leading check so you can use it as a trailing comment if you want to.

Huh, that's an interesting idea. I suppose there's no reason why not to do that. Sure!

**Comment by vjeux** [comment]
I just updated it to `prettier-ignore` and removed the leading check.

```js
  // Incorrectly indented on purpose
      function f</* prettier-ignore */ T    :    B>(
        a : Array  <   number   > // prettier-ignore
      ) {

        call(
          f(         1          )
          // prettier-ignore
        )
      }
```

it works really well!

**Comment by jlongster** [comment]
Haha... /me cringes. Well, unleash the hounds I suppose.

**Comment by sonaye** [comment]
@vjeux how does this work exactly, can you summarize? couldn't find any docs on this. was checking on any #582 related updates.

**Comment by vjeux** [comment]
Oops, forgot to add documentation.

If you put a comment named `prettier-ignore`, it is going to print the raw text of the node it is attached to.

**Comment by karlhorky** [comment]
Would you guys want documentation for this? Where would it go? Just a quick paragraph with example in the readme?

**Comment by vjeux** [comment]
@karlhorky yeah, would be awesome to have it in the README! PR welcome!

**Comment by karlhorky** [comment]
Alright, I'll take a look at this tomorrow.

**Comment by karlhorky** [comment]
Ok, I've opened prettier/prettier#1082.
